### PR TITLE
ore: improve dep hygiene

### PR DESCRIPTION
**Pull request by danhhz**
A number of these deps weren't even used, so remove those.

Luckily, the most heavyweight deps (tokio, etc) seem to form a nice
group, which we pull out into a "network" feature.

Perhaps more controversially, we go further and make all ore deps
optional. This allows ore users that only need the dep-free bits
(notably `cast::CastFrom`) to use them without unnecessary deps. The
number of features this introduces is a unfortunate complexity, so
perhaps a more intermediate solution would be better.

### CODE REVIEW

File: src/ore/src/antichain.rs [hunk 1]
```diff
@@ -78,14 +78,6 @@ impl<T: PartialOrder + Ord + Clone + std::fmt::Debug> AntichainToken<T> {
         let mut new_frontier = Antichain::new();
         new_frontier.extend(frontier.into_iter());
 
-        if !<_ as PartialOrder>::less_equal(&self.antichain.frontier(), &new_frontier.borrow()) {
```
Comment (danhhz) [inline src/ore/src/antichain.rs:81]: As discussed offline, maybe better is to just move this file to the coord package.

File: src/sql-parser/Cargo.toml [hunk 2]
```diff
@@ -10,7 +10,7 @@ publish = false
 itertools = "0.9.0"
 lazy_static = "1.4.0"
 log = "0.4.13"
-ore = { path = "../ore" }
+ore = { path = "../ore", features = ["either"] }
```
Comment (danhhz) [inline src/sql-parser/Cargo.toml:13]: The either stuff appears to only be used in exactly one callsite. I propose we just remove Either entirely from the codebase.

File: src/ore/src/iter.rs [hunk 3]
```diff
@@ -179,40 +177,10 @@ where
     fn chain_one(self, item: Self::Item) -> Chain<Self, Once<Self::Item>> {
         self.chain(iter::once(item))
     }
-
-    /// Converts this iterator to a [`FallibleIterator`].
-    fn fallible<T, E>(self) -> fallible_iterator::Convert<Self>
```
Comment (danhhz) [inline src/ore/src/iter.rs:184]: This wasn't used anywhere. (I think, double checking with CI)

**Comment by benesch** [comment]
Ok, so thoughts!

1. Let's put `bytes` behind the "network" feature too. Despite its general purpose name it is only used by the Tokio ecosystem and is meant for zero copy networking.
2. Optionalizing things to this degree is going to cause more trouble than its worth, IME.  I'd strongly recommend introducing a `network` and `cli` feature, marking them both as default features, and then in the persist crate only specifying `default-features = false`. We also don't have the CI infrastructure for running `cargo test` with various subsets of features, so better if the default is "everything".
3. Let's definitely move antichain.rs to coord.
4. `either` is a very very simple dep that is typically considered to be basically part of the stdlib. Can you just live with that one? The hope is that `display_or` becomes extremely common. It's not at the moment, but there are dozens of places all throughout the codebase where it could be used to good effect. (I think the alternative here should be inlining the definition of `Either` in `ore`, not optionalizing it.)

**Comment by danhhz** [comment]
1) "bytes" is used in fmt, which seems pretty general purpose. It's a bummer for someone that wants the Vec impl of FormatBuffer to have to bring in all the networking stuff. Maybe we can remove the FormatBuffer impl on bytes::BytesMut?
2) Yeah I don't love how many features there are either, so I'm all for simplifying them. Tbh, I'm more convinced by a "complexity of figuring out the name of the feature you need" argument than a testing one. CI will test the combinations we use in practice and that seems fine to me (even if other combos are broken that's easy enough to fix if we ever hit it). ymmv of course
2 con't) When I started this PR, I planned on put all features as default features to keep this as uninvasive as possible, but I swapped it because it turns out that there are 21 deps on ore and 17 of them don't need any of these features turned on. How do you feel about no default features?
3) Will do.
4) I understand Either, but my argument is that it's literally used in one callsite in the entire codebase. I just think, from a practical standpoint, I don't think it's caught on in the way you hoped. That said, I don't feel super strongly, so if you do, we can leave it in

**Comment by danhhz** [comment]
Bummer, it does look like we use the FormatBuffer impl on bytes::BytesMut in src/pgrepr/src/value.rs

File: src/ore/src/lib.rs [hunk 4]
```diff
@@ -25,26 +25,34 @@
...
 
+#[cfg(feature = "timely")]
 pub mod antichain;
 pub mod cast;
+#[cfg(feature = "structopt")]
 pub mod cli;
 pub mod codegen;
 pub mod collections;
 pub mod env;
+#[cfg(feature = "bytes")]
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/ore/src/lib.rs]: Good call! That makes this much more palatable. Will do, yeah

**Review by benesch — APPROVED** [review]
This looks great to me modulo one comment I failed to press "submit" on before, about pushing the conditional compilation of `fmt` down onto the `impl FormatBuffer for Bytes` itself.

**Comment by benesch** [comment]
Ack on your argument for (2), btw.

**Comment by benesch** [comment]
Re `display_or`, let me see if we can't go in the opposite direction. It's still relatively new, and I know there are dozens of places where it could be useful.

**Comment by danhhz** [comment]
Okay looks like I misunderstood the issue with turning everything on by default. It's not so much "do any combinations of features cause the ore crate to not compile" as much as "does each ore user compile on its own". I'm flopping my opinion back to match yours (and my initial one) of everything on by default

**Review by benesch — APPROVED** [review]
Still looks great! Lmk if you want help tracking down that test failure.

**Comment by benesch** [comment]
Oh, I remember now. That explicit dep on `openssl` is actually important, because it enables the `vendored` feature in the `openssl` dep, which is transitively depended upon by `tokio-openssl`.

**Comment by danhhz** [comment]
Thanks for helping with the openssl failure! I'd noticed it but didn't have time to to dig into it yesterday before I stopped for the day. Merging on green.

**Pull request merged**